### ISSUE TITLE
Do not create .ng_deps.dart files when unnecessary

### ISSUE THREAD

**Issue by kegluneq** [angular/angular#1929]
TL;DR: Do not generate `.ng_deps.dart` files for `.dart` files that do not declare any `@Injectable` classes and only import `dart:` libraries.

Currently we generate a `.ng_deps.dart` file for every `.dart` file in the project, but some of these are unnecessary.

#### Changes to `DirectiveProcessor`
In `DirectiveProcessor`, we can avoid creating the `.ng_deps.dart` file iff:

1. The `.dart` file does not define any `@Injectable` classes
2. The `.dart` file does not transitively import any files that define `@Injectable` classes.

We know whether a file has any `@Injectable` classes (which includes subclasses `@Directive` and `@Component`) and what a file's imports are, so while it is difficult to be sure a file does not transitively import anything defining an `@Injectable` class, we can be certain if it only imports `dart:` libraries.

#### Changes to `DirectiveLinker`
In `DirectiveLinker`, we can remove `.ng_deps.dart` files via `Transform#consumeInput` iff:

1. The `.ng_deps.dart` file does not register any `@Injectable` classes ( `NgDeps#registeredTypes.isEmpty`) 
2. None of the file's imports have an associated `.ng_deps.dart` file.

This will not perfectly remove `.ng_deps.dart` files since import cycles could still exist, but it should get pretty close.
